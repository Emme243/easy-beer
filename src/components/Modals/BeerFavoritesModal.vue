<template>
  <div>
    <v-dialog
      v-model="isModalOpen"
      transition="dialog-bottom-transition"
      :overlay-opacity="0.2"
      max-width="400"
      scrollable
    >
      <template #activator="{ on, attrs }">
        <v-badge
          :content="numberOfFavoriteBeers"
          :value="numberOfFavoriteBeers"
          bordered
          color="deep-purple lighten-1"
          overlap
        >
          <v-btn
            :disabled="isFavoriteBeersEmpty"
            color="deep-purple"
            depressed
            small
            v-bind="attrs"
            v-on="on"
          >
            <v-tooltip bottom color="deep-purple">
              <template #activator="{ on, attrs }">
                <v-icon small color="white">fa-solid fa-heart-circle-check</v-icon>
              </template>
              <span>Cervezas favoritas ({{ numberOfFavoriteBeers }})</span>
            </v-tooltip>
          </v-btn>
        </v-badge>
      </template>

      <v-card>
        <v-toolbar dark color="deep-purple">
          <v-btn icon dark @click="isModalOpen = false" data-test-id="close-modal-btn">
            <v-icon small>fa fa-times</v-icon>
          </v-btn>
          <v-toolbar-title>
            <span>Cervezas favoritas ({{ numberOfFavoriteBeers }})</span>
          </v-toolbar-title>
          <v-spacer></v-spacer>
          <v-icon small>fa-solid fa-heart-circle-check</v-icon>
        </v-toolbar>

        <v-card-text class="pt-5">
          <v-alert
            color="deep-purple lighten-1"
            type="info"
            v-if="isFavoriteBeersEmpty"
            data-test-id="no-beers-message"
          >
            No tienes cervezas favoritas por el momento.
          </v-alert>
          <div class="beerFavoritesModal__beerCards" v-else data-test-id="beers-container">
            <BeerInfoCard v-for="beer in favoriteBeersInStore" :key="beer.id" :beer="beer" />
          </div>
        </v-card-text>

        <v-divider />

        <v-card-actions>
          <v-spacer></v-spacer>
          <v-btn
            color="deep-purple"
            depressed
            @click="resetFavoriteBeersInStore"
            :disabled="isFavoriteBeersEmpty"
            data-test-id="reset-favorites-btn"
          >
            <span :class="{ 'white--text': !isFavoriteBeersEmpty }">Vac√≠ar</span>
            <v-icon right color="white">fa fa-eraser</v-icon>
          </v-btn>
          <v-btn
            color="deep-purple"
            outlined
            dark
            depressed
            @click="isModalOpen = false"
            data-test-id="close-modal-cerrar-btn"
          >
            <span>Cerrar</span>
            <v-icon right>fa fa-times</v-icon>
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </div>
</template>

<script>
import BeerInfoCard from '@/components/Cards/BeerInfoCard';
import { mapActions, mapGetters } from 'vuex';

export default {
  name: 'BeerFavoritesModal',
  components: { BeerInfoCard },
  data: () => ({ isModalOpen: false }),
  methods: mapActions({
    resetFavoriteBeersInStore: 'favoriteBeers/resetFavoriteBeers',
  }),
  computed: {
    ...mapGetters({
      favoriteBeersInStore: 'favoriteBeers/getAllFavoriteBeers',
    }),
    numberOfFavoriteBeers() {
      return this.favoriteBeersInStore.length;
    },
    isFavoriteBeersEmpty() {
      return this.numberOfFavoriteBeers === 0;
    },
  },
  watch: {
    favoriteBeersInStore() {
      localStorage.setItem('favoriteBeers', JSON.stringify(this.favoriteBeersInStore));
    },
  },
};
</script>

<style scoped lang="scss">
.beerFavoritesModal__beerCards {
  & > *:not(:last-child) {
    margin-bottom: 20px;
  }
}
</style>
