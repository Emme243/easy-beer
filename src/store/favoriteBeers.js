export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    addBeerToFavorites({ commit }, beer) {
      commit('addBeerToFavorites', beer);
    },
    removeBeerFromFavorites({ commit }, beerId) {
      commit('removeBeerFromFavorites', beerId);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
};
