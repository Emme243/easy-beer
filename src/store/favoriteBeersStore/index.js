export const favoriteBeersStore = {
  namespaced: true,
  state: {
    favoriteBeers: [],
  },
  getters: {
    getAllFavoriteBeers(state) {
      return state.favoriteBeers;
    },
    isBeerInFavorites(state) {
      return id => {
        return state.favoriteBeers.some(beer => beer.id === id);
      };
    },
  },
  mutations: {
    addBeerToFavorites(state, beer) {
      state.favoriteBeers.push(beer);
    },
    removeBeerFromFavorites(state, id) {
      state.favoriteBeers = state.favoriteBeers.filter(beer => beer.id !== id);
    },
    setFavoriteBeers(state, beers) {
      state.favoriteBeers = beers;
    },
    resetFavoriteBeers(state) {
      state.favoriteBeers = [];
    },
  },
  actions: {
    handleBeerInFavorites({ commit, getters }, beer) {
      const isBeerInFavorites = getters.isBeerInFavorites(beer.id);
      if (isBeerInFavorites) commit('removeBeerFromFavorites', beer.id);
      else commit('addBeerToFavorites', beer);
    },
    setFavoriteBeers({ commit }, beers) {
      commit('setFavoriteBeers', beers);
    },
    resetFavoriteBeers({ commit }) {
      commit('resetFavoriteBeers');
    },
  },
};
