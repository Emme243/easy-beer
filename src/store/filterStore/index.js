import { abvRangeStore } from '@/store/filterStore/abvRangeStore';
import { brewedDateStore } from '@/store/filterStore/brewedDateStore';
import { paginationStore } from '@/store/filterStore/paginationStore';
import { searchQueryStore } from '@/store/filterStore/searchQueryStore';

export const filterStore = {
  namespaced: true,
  modules: {
    abvRange: abvRangeStore,
    brewedDate: brewedDateStore,
    page: paginationStore,
    searchQuery: searchQueryStore,
  },
  state: () => ({}),
  getters: {
    filterStore(_, getters) {
      return {
        minAbvValue: getters.minAbvValue,
        maxAbvValue: getters.maxAbvValue,
        initialBrewedMonth: getters.initialBrewedMonth,
        finalBrewedMonth: getters.finalBrewedMonth,
        page: getters.page,
        searchQuery: getters.searchQuery,
      };
    },
    areFiltersWithDefaultValue(_, getters) {
      return (
        getters.isAbvRangeWithDefaultValue &&
        !getters.initialBrewedMonth &&
        !getters.finalBrewedMonth &&
        !getters.searchQuery
      );
    },
    numberOfFiltersApplied(_, getters) {
      let numberOfFiltersApplied = 0;
      if (!getters.isAbvRangeWithDefaultValue) numberOfFiltersApplied++;
      if (getters.initialBrewedMonth || getters.finalBrewedMonth) numberOfFiltersApplied++;
      return numberOfFiltersApplied;
    },
  },
  mutations: {},
  actions: {
    setAllFilters({ dispatch }, filterQuery) {
      dispatch('setPage', filterQuery.page);
      dispatch('setMinAbvValue', filterQuery.minAbvValue);
      dispatch('setMaxAbvValue', filterQuery.maxAbvValue);
      dispatch('setInitialBrewedMonth', filterQuery.initialBrewedMonth);
      dispatch('setFinalBrewedMonth', filterQuery.finalBrewedMonth);
      dispatch('setSearchQuery', filterQuery.searchQuery);
    },
    resetAllFilters({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('setSearchQuery', '');
      dispatch('resetPage');
    },
    resetFiltersInFilterModal({ dispatch }) {
      dispatch('resetAbvRange');
      dispatch('resetBrewedDate');
      dispatch('resetPage');
    },
  },
};
