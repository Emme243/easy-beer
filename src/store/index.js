import { favoriteBeersStore } from '@/store/favoriteBeersStore';
import { filterStore } from '@/store/filterStore';

import Vue from 'vue';
import Vuex from 'vuex';

Vue.use(Vuex);

export default new Vuex.Store({
  modules: {
    filter: filterStore,
    favoriteBeers: favoriteBeersStore,
  },
});
