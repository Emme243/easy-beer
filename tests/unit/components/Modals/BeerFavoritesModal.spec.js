import BeerFavoritesModal from '@/components/Modals/BeerFavoritesModal';
import store from '@/store';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import Vuetify from 'vuetify';
import { beersMock } from '../../mocks/beersMock';

describe('Pruebas en BeerFavoritesModal.spec.js', function () {
  let vuetify;
  const localVue = createLocalVue();

  const mountFunction = options => {
    return shallowMount(BeerFavoritesModal, {
      localVue,
      vuetify,
      store,
      ...options,
    });
  };

  beforeEach(() => {
    vuetify = new Vuetify();
  });

  const beer = beersMock[0];
  const closeModalBtnSelector = '[data-test-id="close-modal-btn"]';
  const closeModalCerrarBtnSelector = '[data-test-id="close-modal-cerrar-btn"]';
  const noBeersMessageSelector = '[data-test-id="no-beers-message"]';
  const beersContainerSelector = '[data-test-id="beers-container"]';
  const resetFavoritesBtnSelector = '[data-test-id="reset-favorites-btn"]';

  it('Debe hacer match con el snapshot', () => {
    const wrapper = mountFunction();
    expect(wrapper.element).toMatchSnapshot();
  });

  it('El bot贸n closeModalBtn debe setear la propiedad isModalOpen en false cuando se haga click', async () => {
    const wrapper = mountFunction({ data: () => ({ isModalOpen: true }) });
    const closeModalBtn = wrapper.find(closeModalBtnSelector);
    expect(closeModalBtn.exists()).toBe(true);
    expect(wrapper.vm.isModalOpen).toBe(true);
    closeModalBtn.vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isModalOpen).toBe(false);
  });

  it('El bot贸n closeModalCerrarBtn debe setear la propiedad isModalOpen en false cuando se haga click', async () => {
    const wrapper = mountFunction({ data: () => ({ isModalOpen: true }) });
    const closeModalCerrarBtn = wrapper.find(closeModalCerrarBtnSelector);
    expect(closeModalCerrarBtn.exists()).toBe(true);
    expect(wrapper.vm.isModalOpen).toBe(true);
    closeModalCerrarBtn.vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(wrapper.vm.isModalOpen).toBe(false);
  });

  it('El mensaje noBeersMessage debe mostrarse cuando no hay bebidas favoritas', async () => {
    const wrapper = mountFunction();
    expect(wrapper.find(noBeersMessageSelector).exists()).toBe(true);
  });

  it('El mensaje noBeersMessage debe ocultarse cuando hay bebidas favoritas', async () => {
    const wrapper = mountFunction({ computed: { favoriteBeersInStore: () => [beer] } });
    expect(wrapper.find(noBeersMessageSelector).exists()).toBe(false);
  });

  it('El contenedor beersContainer debe mostrarse cuando hay bebidas favoritas', async () => {
    const wrapper = mountFunction({ computed: { favoriteBeersInStore: () => [beer] } });
    expect(wrapper.find(beersContainerSelector).exists()).toBe(true);
  });

  it('El contenedor beersContainer debe ocultarse cuando no hay bebidas favoritas', async () => {
    const wrapper = mountFunction();
    expect(wrapper.find(beersContainerSelector).exists()).toBe(false);
  });

  it('El bot贸n resetFavoritesBtn debe llamar al dispatch resetFavoriteBeers cuando se haga click, esto cuando haya cervezas favoritas', async () => {
    store.dispatch = jest.fn();
    const wrapper = mountFunction({ computed: { favoriteBeersInStore: () => [beer] } });
    const resetFavoritesBtn = wrapper.find(resetFavoritesBtnSelector);
    expect(resetFavoritesBtn.exists()).toBe(true);
    resetFavoritesBtn.vm.$emit('click');
    await wrapper.vm.$nextTick();
    expect(store.dispatch).toHaveBeenCalledWith('favoriteBeers/resetFavoriteBeers');
    expect(store.dispatch).toHaveBeenCalledTimes(1);
  });

  it('El bot贸n resetFavoritesBtn debe estar deshabilitado cuando no haya cervezas favoritas', async () => {
    const wrapper = mountFunction();
    const resetFavoritesBtn = wrapper.find(resetFavoritesBtnSelector);
    expect(resetFavoritesBtn.exists()).toBe(true);
    expect(resetFavoritesBtn.attributes().disabled).toBeTruthy();
  });
});
