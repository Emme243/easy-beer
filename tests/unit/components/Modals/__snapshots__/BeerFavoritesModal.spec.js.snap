// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en BeerFavoritesModal.spec.js Debe hacer match con el snapshot 1`] = `
<div>
  <v-dialog-stub
    closedelay="0"
    contentclass=""
    maxwidth="400"
    opendelay="0"
    openonclick="true"
    origin="center center"
    overlayopacity="0.2"
    retainfocus="true"
    scrollable="true"
    transition="dialog-bottom-transition"
  >
     
    <v-card-stub
      loaderheight="4"
      tag="div"
    >
      <v-toolbar-stub
        color="deep-purple"
        dark="true"
        extensionheight="48"
        src=""
        tag="header"
      >
        <v-btn-stub
          activeclass=""
          dark="true"
          data-test-id="close-modal-btn"
          icon="true"
          tag="button"
          type="button"
        >
          <v-icon-stub
            small=""
          >
            fa fa-times
          </v-icon-stub>
        </v-btn-stub>
         
        <v-toolbar-title-stub
          tag="div"
        >
          <span>
            Cervezas favoritas (0)
          </span>
        </v-toolbar-title-stub>
         
        <v-spacer-stub
          tag="div"
        />
         
        <v-icon-stub
          small=""
        >
          fa-solid fa-heart-circle-check
        </v-icon-stub>
      </v-toolbar-stub>
       
      <v-card-text-stub
        class="pt-5"
        tag="div"
      >
        <v-alert-stub
          closeicon="$cancel"
          closelabel="$vuetify.close"
          color="deep-purple lighten-1"
          data-test-id="no-beers-message"
          icon=""
          tag="div"
          type="info"
          value="true"
        >
          
          No tienes cervezas favoritas por el momento.
        
        </v-alert-stub>
      </v-card-text-stub>
       
      <v-divider-stub />
       
      <v-card-actions-stub
        tag="div"
      >
        <v-spacer-stub
          tag="div"
        />
         
        <v-btn-stub
          activeclass=""
          color="deep-purple"
          data-test-id="reset-favorites-btn"
          depressed="true"
          disabled="true"
          tag="button"
          type="button"
        >
          <span
            class=""
          >
            Vac√≠ar
          </span>
           
          <v-icon-stub
            color="white"
            right=""
          >
            fa fa-eraser
          </v-icon-stub>
        </v-btn-stub>
         
        <v-btn-stub
          activeclass=""
          color="deep-purple"
          dark="true"
          data-test-id="close-modal-cerrar-btn"
          depressed="true"
          outlined="true"
          tag="button"
          type="button"
        >
          <span>
            Cerrar
          </span>
           
          <v-icon-stub
            right=""
          >
            fa fa-times
          </v-icon-stub>
        </v-btn-stub>
      </v-card-actions-stub>
    </v-card-stub>
  </v-dialog-stub>
</div>
`;
